### PR TITLE
RavenDB-6326 related implementation

### PR DESCRIPTION
**Pull request by myarichuk**
also add commands GetConflicts and ResolveConflicts (for use in tests and in general for usage in the new client)

### CODE REVIEW

File: src/Raven.Client/Constants.cs [hunk 1]
```diff
@@ -200,6 +202,8 @@ private Encryption()
 
             public class Replication
             {
+                public const string ChangeVector = "@change-vector";
```
Comment (ayende) [inline src/Raven.Client/Constants.cs:205]: Why have both `const`?

File: src/Raven.Client/Documents/Commands/ResolveConflictCommand.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        public ResolveConflictCommand(string id,ChangeVectorEntry[] changeVector, JsonOperationContext context)
+        {
+            _id = id;
+            _changeVector = changeVector;
+            _context = context;
+        }
+
+        public override HttpRequestMessage CreateRequest(ServerNode node, out string url)
+        {
+            url = $"{node.Url}/databases/{node.Database}/replication/conflicts?docId={_id}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/ResolveConflictCommand.cs:31]: What is this endpoint? And why do we need a special command?

File: src/Raven.Client/Documents/Replication/Messages/ReplicationMessageReply.cs [hunk 3]
```diff
@@ -67,5 +80,14 @@ public DynamicJsonValue ToJson()
                 [nameof(Etag)] = Etag
             };
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (ayende) [inline src/Raven.Client/Documents/Replication/Messages/ReplicationMessageReply.cs:84]: Stop calling those, they don't do what you think they do

File: src/Raven.Client/Documents/Replication/Messages/ReplicationMessageReply.cs [hunk 4]
```diff
@@ -67,5 +80,14 @@ public DynamicJsonValue ToJson()
...
+        public bool CanConvert() => true;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (ayende) [inline src/Raven.Client/Documents/Replication/Messages/ReplicationMessageReply.cs:87]: Remove

File: src/Raven.Client/Extensions/BlittableJsonExtensions.cs [hunk 5]
```diff
@@ -1,10 +1,46 @@
...
     internal static class BlittableJsonExtensions
     {
+        public static ChangeVectorEntry[] ToVector(this BlittableJsonReaderArray vectorJson)
+        {
+            var result = new ChangeVectorEntry[vectorJson.Length];
+            int iter = 0;
+            foreach (BlittableJsonReaderObject entryJson in vectorJson)
+            {
+                if (!entryJson.TryGet(nameof(ChangeVectorEntry.DbId), out result[iter].DbId))
+                    throw new InvalidDataException();
```
Comment (ayende) [inline src/Raven.Client/Extensions/BlittableJsonExtensions.cs:18]: Explain what is the error

File: src/Raven.Client/Extensions/BlittableJsonExtensions.cs [hunk 6]
```diff
@@ -1,10 +1,46 @@
...
+                if (!entryJson.TryGet(nameof(ChangeVectorEntry.Etag), out result[iter].Etag))
+                    throw new InvalidDataException();
```
Comment (ayende) [inline src/Raven.Client/Extensions/BlittableJsonExtensions.cs:20]: Explain what is the error and where it happens.

File: src/Raven.Client/Json/MetadataAsDictionary.cs [hunk 7]
```diff
@@ -34,7 +35,10 @@ public void Init()
                     return _metadata[key];
                 object value;
                 if (_source.TryGetMember(key, out value))
-                    return value.ToString();
+                {
+                    var blittableArray = value as BlittableJsonReaderArray;
+                    return blittableArray != null ? blittableArray.ToJsonString() : value.ToString();
```
Comment (ayende) [inline src/Raven.Client/Json/MetadataAsDictionary.cs:40]: Why would this ever be called?

File: src/Raven.Client/Extensions/BlittableJsonExtensions.cs [hunk 8]
```diff
@@ -1,10 +1,46 @@
...
+
+                iter++;
+            }
+            return result;
+        }
+
+        public static string ToJsonString(this BlittableJsonReaderArray vectorJson)
```
Comment (ayende) [inline src/Raven.Client/Extensions/BlittableJsonExtensions.cs:27]: Remove this, we never want this to happen.

File: src/Raven.Client/Json/BlittableJsonWriter.cs [hunk 9]
```diff
@@ -63,15 +64,47 @@ private void WriteMetadata()
...
                     var propertyDetails = new BlittableJsonReaderObject.PropertyDetails();
                     _documentInfo.Metadata.GetPropertyByIndex(id, ref propertyDetails);
-
                     _manualBlittalbeJsonDocumentBuilder.WritePropertyName(propertyDetails.Name);
+
                     switch (propertyDetails.Token)
                     {
+                        //in this case it can only be change vector (since it is the only array in the metadata)
+                        //not sure how viable/maintenable this is.
+                        case BlittableJsonToken.StartArray | BlittableJsonToken.OffsetSizeByte:
```
Comment (ayende) [inline src/Raven.Client/Json/BlittableJsonWriter.cs:73]: We _never_ do a check with the offset, you need to mask them at the switch

File: src/Raven.Client/Json/BlittableJsonWriter.cs [hunk 10]
```diff
@@ -63,15 +64,47 @@ private void WriteMetadata()
                 {
                     var propertyDetails = new BlittableJsonReaderObject.PropertyDetails();
                     _documentInfo.Metadata.GetPropertyByIndex(id, ref propertyDetails);
-
                     _manualBlittalbeJsonDocumentBuilder.WritePropertyName(propertyDetails.Name);
+
                     switch (propertyDetails.Token)
                     {
+                        //in this case it can only be change vector (since it is the only array in the metadata)
+                        //not sure how viable/maintenable this is.
```
Comment (ayende) [inline src/Raven.Client/Json/BlittableJsonWriter.cs:72]: Add a check for the property name

File: src/Raven.Client/Json/MetadataAsDictionary.cs [hunk 11]
```diff
@@ -65,7 +69,10 @@ public ICollection<string> Values
                 {
                     var propDetails = new BlittableJsonReaderObject.PropertyDetails();
                     _source.GetPropertyByIndex(prop, ref propDetails);
-                    values.Add(propDetails.Value.ToString());
+                    var blittableArray = propDetails.Value as BlittableJsonReaderArray;
+                    values.Add(blittableArray != null ? 
+                                    blittableArray.ToJsonString() :
```
Comment (ayende) [inline src/Raven.Client/Json/MetadataAsDictionary.cs:74]: remove this

File: src/Raven.Server/Documents/Document.cs [hunk 12]
```diff
@@ -64,7 +64,7 @@ public void EnsureMetadata(float? indexScore = null)
 
             mutatedMetadata[Constants.Documents.Metadata.Etag] = Etag;
             mutatedMetadata[Constants.Documents.Metadata.Id] = Key;
-
+            mutatedMetadata[Constants.Documents.Metadata.ChangeVector] = ChangeVector;
```
Comment (ayende) [inline src/Raven.Server/Documents/Document.cs:67]: we never save the change vector back

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 13]
```diff
@@ -1131,17 +1132,24 @@ private static void ThrowOnDocumentConflict(DocumentsOperationContext context, S
...
             {
-                var changeVectors = new List<ChangeVectorEntry[]>();
+                var conflictRecords = new List<GetConflictsResult.Conflict>();
                 foreach (var conflict in conflicts)
-                    changeVectors.Add(conflict.ChangeVector);
+                {
+                    conflictRecords.Add(new GetConflictsResult.Conflict
+                    {
+                        Doc = conflict.Doc,
+                        Key = conflict.LoweredKey,
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs:1141]: Why `LoweredKey`, and make sure that we don't hold to memory here from another context

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 14]
```diff
@@ -1160,6 +1168,52 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
                 etagTree.Add(LastEtagSlice, etagSlice);
         }
 
+        public void AddTombstoneIfNoDocument(
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs:1171]: Remove this

File: src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs [hunk 15]
```diff
@@ -46,6 +51,89 @@ public Task GetAllTombstones()
             return Task.CompletedTask;
         }
 
+        [RavenAction("/databases/*/replication/conflicts", "PUT", "/databases/{databaseName:string}/replication/conflicts?docId={documentId:string}")]
+        public Task SolveReplicationConflictById()
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs:55]: Remove this

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by myarichuk** [comment]
Closing for now, will reopen after implementing changes

**Pull request closed without merging**